### PR TITLE
Parquet dataloader with fairseq2 primitives 

### PR DESCRIPTION
**Pull request by artemru**
**What does this PR do? Please describe:**
We introduce a general purpose Parquet dataloader based on fairseq2 DataPipeline API.
Parquet interface is done via [pyarrow](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetDataset.html#pyarrow.parquet.ParquetDataset)

As a usage example, we provide an ASR specific data loader built on top of the generic parquet dataloader.

**Does your PR introduce any breaking changes? If yes, please list them:**
This contribution is completely independent and does not modify other components.

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -46,6 +46,8 @@
         "fairseq2n" + fairseq2n_version_spec,
         "jiwer~=3.0",
         "numpy~=1.23",
+        "pyarrow>=13.0.0",
+        "pandas>=2.0.0",
```
Comment (artemru) [inline setup.py:50]: to avoid this extra deps that could be heavy, we could provide specific installation with `pip install fairseq2[arrow]`.

File: src/fairseq2/utils/parquet_dataloader.py [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    def _sanitize_dict(table: pa.Table) -> pa.Table:
+        for i, (name, type) in enumerate(zip(table.schema.names, table.schema.types)):
+            if pa.types.is_dictionary(type):
+                ca = table[name].cast(type.value_type)
+                table = table.remove_column(i)
+                table = table.append_column(name, ca)
+        return table
+
+    @staticmethod
+    def compute_length_splits(
```
Comment (gwenzek) [inline src/fairseq2/utils/parquet_dataloader.py:349]: what's the benefice of having this logic here and not in the pipeline like we do with the other fairseq2 dataloader ?

File: tests/unit/data/data_pipeline/test_parquet_dataloader.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    pq.write_to_dataset(
+        table,
+        path,
+        partition_cols=["part_key"] if n_partitions > 0 else None,
+        existing_data_behavior="delete_matching",
+        **{"row_group_size": 110},
+    )
+
+
+class TestParquetDataloader(unittest.TestCase):
```
Comment (gwenzek) [inline tests/unit/data/data_pipeline/test_parquet_dataloader.py:75]: can you use pytest please ?
to get a temporary dir with pytest you just make a function with a `tmp_path: Path` as argument:

```py
def test_simple_dataload(tmp_path: Path):
        config = ParquetBasicDataloaderConfig(
            parquet_path=tmp_path / "test.pq", ...)
```

File: tests/unit/data/data_pipeline/test_parquet_dataloader.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+            nb_producers=2,
+            seed=333,
+        )
+        pbdl = ParquetBasicDataLoader(config)
+        res: tp.List[pd.DataFrame] = list(iter(pbdl))
+
+        for x in res:
+            self.assertIsInstance(x, pa.Table)
+
+        self.assertEqual(
```
Comment (gwenzek) [inline tests/unit/data/data_pipeline/test_parquet_dataloader.py:110]: in pytest this just become `assert ... == ...`

**Review by gwenzek — COMMENTED** [review]
The individual function implementation looks good, but the overall shape of the code should be reversed.

In fairseq2.data where we try to expose stuff to the user, in particular the pipeline.
Here you're making a bunch of assumptions on how people should read parquet data, and it doesn't play well with the rest of fairseq2.

What I was suggesting is that you expose:

* functions to find and filter list of parquet files (fragments) 
* one function to load a fragment and yield the individual rows
* one function to convert a row into a friendly type for fairseq2.data

File: src/fairseq2/utils/parquet_dataloader.py [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+                splits.append(argsort_ind[ptr : (i - 1)])
+                ptr = i - 1
+        if (
+            length <= max_tokens
+        ):  # we drop the last iteration if it results in a batch greater than max_tokens
+            splits.append(argsort_ind[ptr:])
+        return splits
+
+    @staticmethod
+    def compute_length(pa_array: pa.Array) -> npt.NDArray[np.int32]:
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/parquet_dataloader.py:378]: why ?

File: src/fairseq2/utils/parquet_dataloader.py [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+        }
+
+    if pa.types.is_nested(arr_type):
+        # TODO: deal with arr = [[{'a': 1}, {'a': 2}]]
+        pass
+
+    if strict:
+        raise NotImplementedError(f"{arr_type} cannot be converted to torch.Tensor")
+    else:
+        return arr.to_pandas()
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/parquet_dataloader.py:225]: in case, where I dont find a way to convert to torch dtype, I prefer to return pandas dtype (which is more commonly used), yet I could raise or keep pyarrow.Array...

**Comment by artemru** [comment]
Thanks for these notes ! I may rearrange the code a bit to expose intermediate datapipeline checkpoints.
> * one function to convert a row into a friendly type for fairseq2.data
Please, note that I'm not converting a single row but the whole batch and that's what exactly `from_pyarrow_to_torch_tensor` function does !

File: src/fairseq2/utils/parquet_dataloader.py (hunk 2 above)
Comment (artemru) [reply src/fairseq2/utils/parquet_dataloader.py:348]: yes, indeed a good point ! can you please show me how to realize this logic with fairseq2 primitives ?

File: src/fairseq2/utils/parquet_dataloader.py [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+        seed = seed if seed is not None else self.config.seed
+        np_rs = np.random.RandomState(
+            hash((seed, epoch)) % 2**32
+        )  # TODO: use stable hashing instead python
+        if self.config.shuffle:
+            all_shuffled_fragments = list(np_rs.permutation(self._all_fragments))
+        else:
+            all_shuffled_fragments = self._all_fragments
+
+        pipeline_builder = (
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/parquet_dataloader.py:470]: I dont see exactly what you mean ? My first goal was to create a parquet dataloader pipeline as a whole (somehow config stats to be similar to what [petastorm](https://github.com/uber/petastorm/blob/master/petastorm/reader.py) does) ... of course, I could expose more tools to assemble it, but for now I dont see a real use case - and in the future, once we got one, I'm open to a refactor.

File: src/fairseq2/utils/parquet_dataloader.py [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+
+    def build_epoch_iterator_pipeline(
+        self, seed: tp.Optional[int] = None, epoch: int = 0
+    ) -> DataPipelineBuilder:
+        seed = seed if seed is not None else self.config.seed
+        np_rs = np.random.RandomState(
+            hash((seed, epoch)) % 2**32
+        )  # TODO: use stable hashing instead python
+        if self.config.shuffle:
+            all_shuffled_fragments = list(np_rs.permutation(self._all_fragments))
```
[the remark this answers is not in the document]
Comment (gwenzek) [reply src/fairseq2/utils/parquet_dataloader.py:482]: fairseq2n uses torch RNG, so  you use torch.seed to set the seed.

File: src/fairseq2/utils/parquet_dataloader.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gwenzek) [reply src/fairseq2/utils/parquet_dataloader.py:486]: since fairseq2 is a library, we should expose the bits needed to build a parquet dataloader,  then the pipeline code itself can go in stopes.

File: src/fairseq2/utils/parquet_dataloader.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gwenzek) [reply src/fairseq2/utils/parquet_dataloader.py:393]: if you want them to be "private" you should prepend the name with "_". This is orthogonal to staticmethod vs function.

**Comment by gwenzek** [comment]
> one function to convert a row into a friendly type for fairseq2.data
Please, note that I'm not converting a single row but the whole batch and that's what exactly from_pyarrow_to_torch_tensor function does !

Yes, but I'm not sure I like it. What's the point ? It seem to force the batching of your model to be the same than the row group of your parquet files. What if you want to change the granularity ? You'll have to split the row groups in row at some point. That's why I'm saying this. The fact that you're using row group to group the reads is great, but that's not how the data should be returned to the user.

File: src/fairseq2/utils/parquet_dataloader.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/parquet_dataloader.py:482]: ok, I'll take this !

**Comment by artemru** [comment]
@gwenzek, @cbalioglu what we could do to avoid using `_TableWrapper` (so that fairseq2.data does not try to iterate over pa.Table/pd.DataFrame columns when we call `.map`) ?

File: src/fairseq2/utils/parquet_dataloader.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/parquet_dataloader.py:486]: For me, splitting a simple dataloader class between two libraries would not make the live of end users easier...
I'll try to expose as much methods as I can so they could be reused in the future. 
If you do have some specific usecase in mind I could try to consider them as well.

File: src/fairseq2/utils/parquet_dataloader.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/parquet_dataloader.py]: done

File: tests/unit/data/data_pipeline/test_parquet_dataloader.py (hunk 4 above)
Comment (artemru) [reply tests/unit/data/data_pipeline/test_parquet_dataloader.py:110]: i'm staying in UnitTest conventions since I started to use them

Comment (gwenzek) [reply tests/unit/data/data_pipeline/test_parquet_dataloader.py:110]: since you're trying to merge a PR to fairseq2, it would be better to follow the style and conventions used there.

**Comment by gwenzek** [comment]
> what we could do to avoid using _TableWrapper (so that fairseq2.data does not try to iterate over pa.Table/pd.DataFrame columns when we call .map) ?

To be more explicit here, what happens is that in fairseq2n we call `PySequence_Check` to check if the received object is a sequence, and then convert it to a `std::vector`. The problem is that `pa.Table` passes the check but fails when being iterated over.
I'm not sure there is a good fix on our side, `pa.Table` shouldn't look like a sequence if it can't be used like one.

**Comment by gwenzek** [comment]
see eg https://github.com/fairinternal/fairseq2/commit/9ea7c3bbc64e0ca43d754660ea2cfee6c96dbc0c
for a simpler API.

The main issue I see is that I'd like to returns individual rows from parquet file, but `yield_from` don't support python generators so it's a bit limiting here.

**Comment by gwenzek** [comment]
see https://github.com/fairinternal/fairseq2/pull/549/commits/6b2b0181acaa1c9aea1117ef51b83582d62c8a5a for a "read_rows" implementation.

I haven't done perf benchmark, I suspect "read_rows" and "read_fragments" could become a bottleneck, but they could be rewritten in C++

File: src/fairseq2/utils/parquet_dataloader.py [hunk 9]
```diff
@@ -124,6 +124,12 @@ class ParquetBasicDataloaderConfig:
     Whether pyarrow should use its internal parallelism threads to read the parquet part.
     Since we rely on the external parallelism, this param is tuned off.
     """
+    filesystem: tp.Optional[pa.fs.FileSystem] = None
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/parquet_dataloader.py:127]: ok, I see what you mean, let me try to refactor the code to split this dependency. 
That said it'll be impossible to get rid of the break-changing modifications in pyarrow anyway.

**Comment by cbalioglu** [comment]
I am going to have a more detailed review tomorrow, but after having an initial quick check, I mostly share the same sentiment as Guillaume. Although the ParquetDataLoader API is nice and tidy for our specific use case in stopes and seamless_common, I feel that it is out-of-place in fairseq2 and conflicts with the philosophy of fairseq2 (mostly following the Unix philosophy of atomic APIs that can be easily composed together). The ultimate API I would envision would look like this:

```python
pipeline = from_parquet("<path/to/parquet_file", columns=[..]).shard(..).map(..).shuffle()...and_return()
```

Essentially a `from_parquet` op that just reads a parquet file row-by-row and does nothing else beyond that. If that is not possible with the existing Python APIs, my suggestion would be to expose just the fundamental APIs (the ones Guillaume listed), and offload the rest of the logic to seamless_common; with maybe a recipe in fairseq2 to demonstrate how they can be composed together to efficiently read a Parquet file.

**Review by cbalioglu — COMMENTED** [review]
I think before diving into an actual implementation review, we should decide on the right approach on how we want to shape up this API. As I mentioned, to me the majority of this API feels quite monolithic and specific to Seamless and our team's research infra. Since fairseq2 aims to be a general purpose "backend" library, we should ideally cherry pick the low-level parquet specific functions in this PR (e.g. reading fragments, converting a parquet row to a fairseq2 "data") that do the real integration work with Arrow, and move the pipeline building and other boilerplate parts to stopes or seamless_common. This is the approach that we have used in seamless_communication and SONAR and so far worked pretty well.

We can of course have a simple recipe under the recipes directory that demonstrates how a very simple data pipeline can be build using the new low-level parquet ops.

File: src/fairseq2/utils/parquet_dataloader.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/parquet_dataloader.py]: I just pushed the refactored code exposing with the lower level primitives.

File: src/fairseq2/utils/parquet_tools.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        return arr
+
+
+def pyarrow_table_to_torch_dict(tt: pa.Table, strict: bool = True) -> NestedDict:
+    return {
+        col: from_pyarrow_to_torch_tensor(tt[col], strict) for col in tt.column_names
+    }
+
+
+def batch_collater(
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/parquet_tools.py:99]: I should be able to use Collater...

**Comment by artemru** [comment]
> I am going to have a more detailed review tomorrow, but after having an initial quick check, I mostly share the same sentiment as Guillaume. Although the ParquetDataLoader API is nice and tidy for our specific use case in stopes and seamless_common, I feel that it is out-of-place in fairseq2 and conflicts with the philosophy of fairseq2 (mostly following the Unix philosophy of atomic APIs that can be easily composed together). The ultimate API I would envision would look like this:
> 
> ```python
> pipeline = from_parquet("<path/to/parquet_file", columns=[..]).shard(..).map(..).shuffle()...and_return()
> ```
> 
> Essentially a `from_parquet` op that just reads a parquet file row-by-row and does nothing else beyond that. If that is not possible with the existing Python APIs, my suggestion would be to expose just the fundamental APIs (the ones Guillaume listed), and offload the rest of the logic to seamless_common; with maybe a recipe in fairseq2 to demonstrate how they can be composed together to efficiently read a Parquet file.

I'm not 100% about the right API for `from_parquet`, but for me it's not row-by-row (since it's the opposite of pyarrow column oriented logic). I would say that `from_parquet` should naturally return `pa.Tables` corresponding to loaded row groups. Split&Rebatch operations does not necessary come by row-by-row logic as you can see in the current implentation.

I'm ok to move ASR specific dataloader to sealess_communication, but I would like to keep the parquet primitives as well as a building a basic parquet dataloader here as an example of usage. wdyt ?

**Comment by artemru** [comment]
> > what we could do to avoid using _TableWrapper (so that fairseq2.data does not try to iterate over pa.Table/pd.DataFrame columns when we call .map) ?
> 
> To be more explicit here, what happens is that in fairseq2n we call `PySequence_Check` to check if the received object is a sequence, and then convert it to a `std::vector`. The problem is that `pa.Table` passes the check but fails when being iterated over. I'm not sure there is a good fix on our side, `pa.Table` shouldn't look like a sequence if it can't be used like one.

I think we should be able to transport arbitrary python object inside fairseq2 data pipelines without unnecessary cast to the sequences... or a at least to turn off this implicit behavior.

File: src/fairseq2/utils/asr_parquet_dataloader.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/utils/asr_parquet_dataloader.py:1]: Agree on that one - it's very seamless_common specific indeed !

**Review by cbalioglu — APPROVED** [review]
